django/db/models/expressions.py

```diff
@@ -283,18 +284,18 @@ def _output_field_or_none(self):
 
     def _resolve_output_field(self):
         """
-        Attempt to infer the output type of the expression. If the output
-        fields of all source fields match then, simply infer the same type
-        here. This isn't always correct, but it makes sense most of the time.
+        Attempt to infer the output type of the expression.
 
-        Consider the difference between `2 + 2` and `2 / 3`. Inferring
-        the type here is a convenience for the common case. The user should
-        supply their own output_field with more complex computations.
+        As a guess, if the output fields of all source fields match then simply
+        infer the same type here.
 
         If a source's output field resolves to None, exclude it from this check.
         If all sources are None, then an error is raised higher up the stack in
         the output_field property.
         """
+        # This guess is mostly a bad idea, but there is quite a lot of code
+        # (especially 3rd party Func subclasses) that depend on it, we'd need a
+        # deprecation path to fix it.
```

Comment:
Completely agree that this was a bad idea, it shouldn't be too hard to deprecate and I assume it wouldn't be too disruptive if we defined explicit `output_field` (or proper logic) for all built-in expressions.

This is something ticket-30446/#11359 forced us to do in a few instances already anyway (e.g. ticket-33114)